python-package/xgboost/spark/core.py

```diff
@@ -0,0 +1,877 @@
+                f"or you need increase spark cluster workers."
+            )
+
+        if self._repartition_needed(dataset):
+            dataset = dataset.withColumn(
+                "values", col("values").cast(ArrayType(FloatType()))
+            )
+            dataset = dataset.repartition(num_workers)
+        train_params = self._get_distributed_config(dataset, fit_params)
+
+        def _train_booster(pandas_df_iter):
+            """
+            Takes in an RDD partition and outputs a booster for that partition after going through
+            the Rabit Ring protocol
+            """
+            from pyspark import BarrierTaskContext
+
+            context = BarrierTaskContext.get()
+
+            dtrain, dval = None, []
+            if has_validation:
+                dtrain, dval = convert_partition_data_to_dmatrix(
+                    pandas_df_iter, has_weight, has_validation
+                )
+                dval = [(dtrain, "training"), (dval, "validation")]
+            else:
+                dtrain = convert_partition_data_to_dmatrix(
+                    pandas_df_iter, has_weight, has_validation
+                )
+
+            booster_params, kwargs_params = self._get_dist_booster_params(train_params)
+            context.barrier()
+            _rabit_args = ""
+            if context.partitionId() == 0:
+                _rabit_args = str(_get_rabit_args(context, num_workers))
+
+            messages = context.allGather(message=str(_rabit_args))
+            _rabit_args = _get_args_from_message_list(messages)
+            evals_result = {}
+            with RabitContext(_rabit_args, context):
+                booster = worker_train(
+                    params=booster_params,
+                    dtrain=dtrain,
+                    evals=dval,
+                    evals_result=evals_result,
+                    **kwargs_params,
+                )
+            context.barrier()
+
+            if context.partitionId() == 0:
+                yield pd.DataFrame(data={"booster_bytes": [cloudpickle.dumps(booster)]})
+
+        result_ser_booster = (
+            dataset.mapInPandas(_train_booster, schema="booster_bytes binary")
+            .rdd.barrier()
+            .mapPartitions(lambda x: x)
+            .collect()[0][0]
+        )
+        result_xgb_model = self._convert_to_model(cloudpickle.loads(result_ser_booster))
+        return self._copyValues(self._create_pyspark_model(result_xgb_model))
+
+    def _fit(self, dataset):
+        self._validate_params()
+        # Unwrap the VectorUDT type column "feature" to 4 primitive columns:
+        # ['features.type', 'features.size', 'features.indices', 'features.values']
+        features_col = col(self.getOrDefault(self.featuresCol))
+        label_col = col(self.getOrDefault(self.labelCol)).alias("label")
+        features_array_col = vector_to_array(features_col, dtype="float32").alias(
```

Comment:
Hi @WeichenXu123, I was thinking the internal of this PR is actually using array col for features by converting the vector to the array, which means users must vectorize their features first, and then de-vectorize by this PR, looks like it's a little bit redundant. How about considering accepting any type of feature col. eg,

1. feature col is array type, we don't need to to anything
2. feature col is vector type, we just call vector_to_array
3. feature col is an array of numeric cols names (AKA, column-wise), we can create the array col by "functions.array"

Reply:
yes make sense.
I will improve later.

Reply:
> feature col is array type, we don't need to to anything
Supported.
> feature col is vector type, we just call vector_to_array
Supported.
> feature col is an array of numeric cols names (AKA, column-wise), we can create the array col by "functions.array"
On second thought, I suggest user to use `pyspark.ml.transform.VectorAssembler` to assemble these columns into a vector column. This is a common patten in pyspark ML.
User can create a pyspark ML pipeline contains transforming stages (do assembling and other tranformations) and xgboost estimator stage.


About @trivialfis 's proposal:
> Prediction doesn't need data concatenation. This can also remove specializations in data processing procedures.

This I also suggest to use `pyspark.ml.transform.VectorAssembler`, and create pyspark ML pipeline so that it can ensure applying the same assembling transformation during training and prediction.